### PR TITLE
Add OCR feature

### PR DESCRIPTION
**Pull request by jlduragos**
## Issue

#153 Capture and return text from image using OCR

## Changes

1. Add a way to get the text from scanned image.
2. Updated IDPassManager as it's already deprecated and some classes are missing. 
3. Updated main view to have better symmetry.
4. Add OCROptions class for plugins to be able to communicate thru the SmartScanner app.
5. Link OCROptions config values to SmartScanner & OCR related screens.
6. Added image extension to check if the image being analyzed by OCR is blurry.
7. Updated the target are UI for MRZ and OCR. 
8. Created and OCRChecker class to check if OCR results are matching the given regex.

### CODE REVIEW

File: core-lib/src/main/java/org/idpass/smartscanner/lib/idpasslite/IDPassManager.kt
Comment (jeremi) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/idpasslite/IDPassManager.kt:1]: Why is this file changed for the OCR feature?

File: core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+import org.idpass.smartscanner.lib.utils.extension.cacheImagePath
+import org.idpass.smartscanner.lib.utils.extension.cacheImageToLocal
+import org.idpass.smartscanner.lib.utils.extension.cropCenter
+import org.idpass.smartscanner.lib.utils.extension.encodeBase64
+import org.idpass.smartscanner.lib.utils.extension.isImageBlur
+import org.idpass.smartscanner.lib.utils.extension.setBrightness
+import org.idpass.smartscanner.lib.utils.extension.setContrast
+import java.io.File
+
+open class OCRAnalyzer(
```
Comment (jeremi) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt:34]: Is this generic OCR? or is it for a specific type of document?
Is regex optional?
What is the country for?

File: core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+        Log.d(SmartScannerActivity.TAG, "Success from OCR")
+        Log.d(SmartScannerActivity.TAG, "value: $result")
+        data.putExtra(SmartScannerActivity.SCANNER_IMAGE_TYPE, imageResultType)
+        data.putExtra(SmartScannerActivity.SCANNER_RESULT, result)
+        data.putExtra(ScannerConstants.MODE, mode)
+        activity.setResult(Activity.RESULT_OK, data)
+        activity.finish()
+    }
+
+    private fun sendBundleResult(barcodeResult: OCRResult? = null) {
```
Comment (jeremi) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt]: Why is it called barcode?

**Review by jeremi — CHANGES_REQUESTED** [review]

File: core-lib/src/main/java/org/idpass/smartscanner/lib/idpasslite/IDPassManager.kt
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/idpasslite/IDPassManager.kt:1]: When I try to change the smartscanner-api aar the IDPassLiteHelper class is already missing. I think there was a cached version on the previous aar. I tried tracking the class but it's also no longer in the library https://github.com/idpass/idpass-lite-java.

File: core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt (hunk 2 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt]: I will fix this one.

Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt]: Fixed

File: core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt (hunk 1 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt:34]: OCR is generic.
Regex is optional and by default OCR will scan all texts.
Country was initially added for scope. I will try to check if this is being used. Otherwise this can be removed.

Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt:34]: Country is removed.

File: app/src/main/java/org/newlogic/smartscanner/MainActivity.kt [hunk 3]
```diff
@@ -145,6 +146,33 @@ class MainActivity : AppCompatActivity() {
...
+                    analyzeStart = 1000
+                ),
+                config = Config(
+                    branding = true,
+                    imageResultType = imageType,
+                    label = "",
+                    isManualCapture = false,
+                    orientation = getOrientation(preference),
+                    showOcrGuide = true,
+                    xGuide = 0.5,
```
Comment (haklyray) [inline app/src/main/java/org/newlogic/smartscanner/MainActivity.kt]: default values should be the full image I think

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt [hunk 4]
```diff
@@ -119,6 +120,10 @@ class SmartScannerActivity : BaseActivity(), OnClickListener {
     private var closeButton: View? = null
     private var rectangle: View? = null
     private var rectangleGuide: View? = null
+    private var ocrGuideContainer: View? = null
+    private var widthGuideOCR: View? = null
```
Comment (haklyray) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: nitpick ocr guide or guide ocr

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt [hunk 5]
```diff
@@ -164,21 +169,25 @@ class SmartScannerActivity : BaseActivity(), OnClickListener {
         flashButton = findViewById(R.id.flash_button)
         closeButton = findViewById(R.id.close_button)
         rectangle = findViewById(R.id.rect_image)
-        rectangleGuide = findViewById(R.id.rect_guide)
+        rectangleGuide = findViewById(R.id.scanner_overlay)
+        ocrGuideContainer = findViewById(R.id.guide_layout)
+        widthGuideOCR = findViewById(R.id.container_width)
```
Comment (haklyray) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: same nitpick

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt [hunk 6]
```diff
@@ -292,6 +302,25 @@ class SmartScannerActivity : BaseActivity(), OnClickListener {
                 viewFinder.visibility = VISIBLE
                 barcodeScannerView?.visibility = GONE
             }
+            if (mode == Modes.OCR.value) {
+                analyzer = OCRAnalyzer(
+                    activity = this,
+                    intent = intent,
+                    imageResultType = config?.imageResultType ?: ImageResultType.PATH.value,
+                    analyzeStart = scannerOptions?.ocrOptions?.analyzeStart ?: 0L,
```
Comment (haklyray) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: above the default value is 1000 and here it is 0L

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt [hunk 7]
```diff
@@ -770,27 +843,108 @@ class SmartScannerActivity : BaseActivity(), OnClickListener {
...
-            rectangleGuide?.alpha = 1f;
-            config?.let { conf ->
-                if (conf.widthGuide != 0) {
-                    val nWidth = TypedValue.applyDimension(
-                        TypedValue.COMPLEX_UNIT_DIP, conf.widthGuide.toFloat(), resources.displayMetrics
-                    ).roundToInt()
-                    rectangleGuide?.layoutParams?.width = nWidth
-                }
+            showMRZGuide()
+        } else  if (config != null && config?.showOcrGuide == true) {
```
Comment (haklyray) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: you are testing config twice

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt [hunk 8]
```diff
@@ -770,27 +843,108 @@ class SmartScannerActivity : BaseActivity(), OnClickListener {
...
+                    resources.displayMetrics
+                ).roundToInt()
+                rectangleGuide?.layoutParams?.width = nWidth
+                widthGuideOCR?.layoutParams?.width = nWidth
             }
 
-        } else {
-            rectangleGuide?.alpha = 0f;
+            // if height guide is not by default
+            if (conf.heightGuide != 70) {
```
Comment (haklyray) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: if compare to default value, it would be better to put that into a constant and use it

File: core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt [hunk 9]
```diff
@@ -0,0 +1,254 @@
...
+            val capture = activity.findViewById<View>(R.id.manual_capture)
+            var inputBitmap = bf
+            var inputRot = rotation
+            var rotatedBF = BitmapUtils.rotateImage(bf, rotation)
+
+            capture.setOnClickListener {
+                captured = true
+            }
+
+            if (isShowGuide != null && isShowGuide) {
```
Comment (haklyray) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt]: thats a weird onem what value isshowGuide can take? we cannot check that it is true only? ignoring null and false?

File: core-lib/src/main/java/org/idpass/smartscanner/lib/utils/extension/Image.kt [hunk 10]
```diff
@@ -175,3 +175,32 @@ fun Bitmap.cropCenter() : Bitmap {
         )
     }
 }
+
+/**
+*Checks if the image quality is within the blur threshold.
+*This will help with blurry images that will yield inaccurate results.
+*/
+fun Bitmap.isImageBlur(threshold: Double) : Boolean {
```
Comment (haklyray) [inline core-lib/src/main/java/org/idpass/smartscanner/lib/utils/extension/Image.kt:183]: for QA to check on the performance on low end phone

File: README.md [hunk 11]
```diff
@@ -42,13 +43,14 @@ If you want to build this library from source, instructions to do so can be foun
...
+| Scan Operation | ODK                                            | Non-ODK                                    |
+|----------------|------------------------------------------------|--------------------------------------------|
 | MRZ            | `org.idpass.smartscanner.odk.MRZ_SCAN`         | `org.idpass.smartscanner.MRZ_SCAN`         |
 | NFC            | `org.idpass.smartscanner.odk.NFC_SCAN`         | `org.idpass.smartscanner.NFC_SCAN`         |
 | Barcode        | `org.idpass.smartscanner.odk.BARCODE_SCAN`     | `org.idpass.smartscanner.BARCODE_SCAN`     |
 | QR code        | `org.idpass.smartscanner.odk.QRCODE_SCAN`      | `org.idpass.smartscanner.QRCODE_SCAN`      |
 | ID PASS Lite   | `org.idpass.smartscanner.odk.IDPASS_LITE_SCAN` | `org.idpass.smartscanner.IDPASS_LITE_SCAN` |
+| OCR            | `org.idpass.smartscanner.odk.OCR_SCAN`         | `org.idpass.smartscanner.OCR_SCAN`         |
 
 To perform an operation, create an intent for the desired operation. This example shows how to call an intent to scan an MRZ (ODK):
```
Comment (haklyray) [inline README.md:55]: we need to add somewhere the options that has been added.. like documentation

**Review by haklyray — CHANGES_REQUESTED** [review]

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt (hunk 4 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: Fixed. Changed to guideContainer and guideWidth. We are already using these for both MRZ and OCR.

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt (hunk 5 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: Fixed. Same as above

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt (hunk 6 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: Changed to 0. analyzeStart in MainActivity is the default value if using the OCR in smartscanner. If intent is from other applications, when undefined or null it should be 0

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt (hunk 7 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: Fixed

File: core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt (hunk 9 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/ocr/OCRAnalyzer.kt]: Fixed

File: app/src/main/java/org/newlogic/smartscanner/MainActivity.kt (hunk 3 above)
Comment (jlduragos) [reply app/src/main/java/org/newlogic/smartscanner/MainActivity.kt]: Changed to full image

File: core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt (hunk 8 above)
Comment (jlduragos) [reply core-lib/src/main/java/org/idpass/smartscanner/lib/SmartScannerActivity.kt]: Fixed

**Review by haklyray — APPROVED** [review]
